src/parser.js

```diff
@@ -1480,43 +1473,53 @@ export default class Parser {
     return hrefVariables.length ? hrefVariables : undefined;
   }
 
-  // Create a Refract asset element containing JSON Schema and push into payload
-  pushSchemaAsset(schema, payload, path) {
-    let handledSchema = false;
+  pushBodyAssets(schema, payload, contentType, pushBody) {
+    let jsonSchema;
 
     try {
-      const jsonSchema = convertSchema(schema);
-      const Asset = this.minim.getElementClass('asset');
-      const schemaAsset = new Asset(JSON.stringify(jsonSchema));
-
-      schemaAsset.classes.push('messageBodySchema');
-      schemaAsset.contentType = 'application/schema+json';
-
-      if (this.generateSourceMap) {
-        this.createSourceMap(schemaAsset, path);
-      }
-
-      this.handleExternalDocs(schemaAsset, schema.externalDocs);
-
-      payload.content.push(schemaAsset);
-      handledSchema = true;
+      jsonSchema = convertSchema(schema);
     } catch (exception) {
       this.createAnnotation(
-        annotations.DATA_LOST, path,
+        annotations.DATA_LOST, this.path,
         'Circular references in schema are not yet supported',
       );
+      return;
     }
 
-    if (handledSchema) {
-      try {
-        const generator = new DataStructureGenerator(this.minim);
-        const dataStructure = generator.generateDataStructure(schema);
-        if (dataStructure) {
-          payload.content.push(dataStructure);
-        }
-      } catch (exception) {
-        // TODO: Expose errors once feature is more-complete
+    if (pushBody) {
+      bodyFromSchema(jsonSchema, payload, this, contentType);
+    }
+
+    this.pushSchemaAsset(schema, jsonSchema, payload, this.path);
+    this.pushDataStructureAsset(schema, payload);
+  }
+
+  // Create a Refract asset element containing JSON Schema and push into payload
+  pushSchemaAsset(schema, jsonSchema, payload, path) {
+    const Asset = this.minim.getElementClass('asset');
+    const schemaAsset = new Asset(JSON.stringify(jsonSchema));
```

Comment:
Isn't this the line which throws the circular reference error? It's not getting caught.

Reply:
The circular reference problem / warning will be emitted before we get this far. `convertSchema` will fail and that is caught further up and warning is made and then returned so body nor schema/datastructure asset changes.

Reply:
There ares tests for this (cirlcular-example fixture) to confirm no functionality is broken.